### PR TITLE
Definitely Typescript

### PR DESCRIPTION
**Pull request by deontologician**
This is going to allow us to add and package typescript, I'll probably move modules incrementally.

Typescript 2.0
Webpack 2.1+

Also had to bump babel so we could get it to do es2015 for everything except modules (since webpack 2.0 has tree shaking with es6 module syntax)
### TODO
- [x] port test webpack config over to webpack 2.0
- [x] fix npm scripts to use new webpack command line env options instead of cross-env
- [x] port all utils modules to typescript

---

This change is [Reviewable](https://reviewable.io/reviews/rethinkdb/horizon/741)

### CODE REVIEW

File: client/webpack.config.js [hunk 1]
```diff
@@ -1,40 +1,153 @@
...
+function determinePlugins(options) {
+  const plugins = [
+    new BannerPlugin('__LICENSE__'),
   ]
+  if (options.devBuild) {
+    plugins.push(new DedupePlugin())
+    plugins.push(new OccurrenceOrderPlugin())
+    plugins.push(new UglifyJsPlugin({
+      compress: {
+        screw_ie8: false,
```
Comment (dalanmiller) [inline client/webpack.config.js:180]: Best option name ever.

**Comment by dalanmiller** [comment]
Should we also define our custom Horizon types in this PR and also rip out Joi?

**Comment by deontologician** [comment]
I dont think we have joi in the client now do we?

**Comment by dalanmiller** [comment]
Ahh I assumed we did, cause it's everywhere else.

File: client/webpack.config.js (hunk 1 above)
Comment (vramana) [inline client/webpack.config.js:180]: Totally agreed :100:

**Comment by deontologician** [comment]
UNfortunately, coverage seems broke

**Comment by deontologician** [comment]
Blargh, broke the tests again

**Comment by niieani** [comment]
TypeScript 2.0 final was just released :)

**Pull request closed without merging**